### PR TITLE
[JavaScript] Microsoft 1DS telemetry pipeline created

### LINKED ISSUE #3058: Feedback Controls Telemetry
# Proposal: Adaptive Cards Feedback Controls Telemetry

## Summary

We'll gather data on the feedback controls’ usage, performance, error messages, and reliability in general and across different platforms.
This will serve as a model for preexisting controls and components, guiding our telemetry implementation in the rest of Adaptive Cards.

## Rationale

* We'd be able to measure the usage of the new feedback controls to get a sense of their adoption on each platform
* We'd be able to identify the most common errors and possible bugs, enabling a more data-driven bug fixing process.
* Overall, the Adaptive Cards team would be able to move to a more data-driven development process.

## Scope

| Capability | Priority |
| :---------- | :------- |
| This proposal will allow the Adaptive Cards (AC) team to capture and send back the predefined metrics for the new feedback controls for offline analysis | Must |
| The telemetry implementation will capture and send trace logging and/or errors back to Microsoft with platform and [other relevant card information] | Must |
| This proposal will allow developers to opt-out of telemetry | Must |
| This proposal will allow the AC team to gather data on the time taken to render their controls | Should |
| This proposal will allow external developers to request to send the telemetry data to themselves  | Won't |

## Important Notes

We considered allowing external developers to send the telemetry data to themselves, but that would require further consultation with CELA. We'll need to be conscious of how we communicate our reasons for this. We may give external developers a dashboard for some of the telemetry data in the future, but not now.
We should start with the JS and UWP 1DS telemetry libraries; the mobile SDKs are less supported in this timeframe.

## Open Questions

* How will we differentiate between developers testing out a feature/control and genuine use in practice?
* Will the 1DS team be able to provide consistent support for our timeframe?

### PR DESCRIPTION
**Pull request by utsavlathia**
fixes #3058 

# Telemetry Pipeline in JavaScript
## Summary
A host can choose to use the 1DS telemetry pipeline now that that is set up in the Adaptive Cards code. The Logger in a future PR will make this task intuitive for a developer

## Updated Readme
The readme is updated with the commands that have to be called

## How Verified

1.  A simple unit test was created as part of the testing files for the class 
2. I manually verified that it works by hooking up to the logger and making sure the events sent through
3. I also manually tested edge cases such as if theree was no valueSet or no correlationID being sent through

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/microsoft/AdaptiveCards/pull/3331)

### CODE REVIEW

**Comment by shalinijoshi19** [comment]
Thanks! Just for consistency, maybe rename "Unit TEsts" to "how Verified" and then add a bullet or numbered list around unit tests + manual verification details

File: source/nodejs/README.md [hunk 1]
```diff
@@ -19,9 +19,23 @@ We use lerna to manage package linking and building.
 2. `npm install`
 3. `npx lerna bootstrap`
 
+## Initialize Telemetry
+
```
Comment (shalinijoshi19) [inline source/nodejs/README.md]: "Internal Microsoft 1DS based telemetry provider implementation. "

File: source/nodejs/README.md [hunk 2]
```diff
@@ -19,9 +19,23 @@ We use lerna to manage package linking and building.
...
+We use Microsoft 1DS as our telemetry provider.
```
Comment (shalinijoshi19) [inline source/nodejs/README.md]: >e use Microsoft 1DS as our telemetry provider [](start = 1, length = 45)

More like "A provider that allows for sending telemetry data to the internal Microsoft service endpoints directly"

File: source/nodejs/README.md [hunk 3]
```diff
@@ -19,9 +19,23 @@ We use lerna to manage package linking and building.
...
+
+This is 1DS documentation:
+https://msasg.visualstudio.com/Shared%20Data/_git/1DS.JavaScript?path=%2Fskus%2FanalyticsWeb%2FREADME.md&version=GBmaster&fullScreen=true
```
Comment (shalinijoshi19) [inline source/nodejs/README.md]: Hmm need to check if it's kosher to refer to internal documentation here - we'll need to follow up. They won't work anyways since it's internal links.

File: source/nodejs/adaptivecards/package.json [hunk 4]
```diff
@@ -56,5 +57,8 @@
 			"jsx",
 			"json"
 		]
+	},
+	"dependencies": {
+		"@ms/1ds-analytics-web-js": "^2.0.4"
```
Comment (shalinijoshi19) [inline source/nodejs/adaptivecards/package.json]: Maybe have this be an "optionalDependency" ([https://docs.npmjs.com/files/package.json#optionaldependencies](https://docs.npmjs.com/files/package.json#optionaldependencies)) to avoid breaking installs for the golden path where this provider is not going to be initialized (which at this point is never by default)

File: source/nodejs/adaptivecards/src/logging/Microsoft1DSLogger/Microsoft1DSLogger.ts [hunk 5]
```diff
@@ -0,0 +1,87 @@
+import { ApplicationInsights } from '@ms/1ds-analytics-web-js'; //Imported for 1DS telemetry SDK
```
Comment (shalinijoshi19) [inline source/nodejs/adaptivecards/src/logging/Microsoft1DSLogger/Microsoft1DSLogger.ts]: Follow pattern for pulling in optional dependencies" [https://docs.npmjs.com/files/package.json#optionaldependencies](https://docs.npmjs.com/files/package.json#optionaldependencies)

File: source/nodejs/adaptivecards/src/__tests__/logging/Microsoft1DS.spec.ts [hunk 6]
```diff
@@ -0,0 +1,8 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+import { Microsoft1DSLogger } from "../../logging/Microsoft1DSLogger/Microsoft1DSLogger";
+
+test('Microsoft1DSLogger should be instantiated', () => {
```
Comment (shalinijoshi19) [inline source/nodejs/adaptivecards/src/__tests__/logging/Microsoft1DS.spec.ts]: >should [](start = 25, length = 6)

*can*

**Comment by shalinijoshi19** [comment]
> 

is this to be checked in? or will it be pulled in automatically as part of the npm install ? 

---
Refers to: source/nodejs/adaptivecards/src/logging/Microsoft1DSLogger/ms.analytics-web-2.min.js:8 in af77e71. [](commit_id = af77e7121a4f43a6da4c7ccce9e37f1e2dc28c8e, deletion_comment = False)

**Comment by utsavlathia** [comment]
> is this to be checked in? or will it be pulled in automatically as part of the npm install ?
> 
> Refers to: source/nodejs/adaptivecards/src/logging/Microsoft1DSLogger/ms.analytics-web-2.min.js:8 in af77e71. [](commit_id = af77e71, deletion_comment = False)

As far as I'm aware this has to be checked in

File: source/nodejs/adaptivecards/package.json (hunk 4 above)
Comment (utsavlathia) [reply source/nodejs/adaptivecards/package.json]: Good idea, I think this may fix the issue we are having

File: source/nodejs/adaptivecards/src/logging/Microsoft1DSLogger/Microsoft1DSLogger.ts [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+
+/**
+ * The purpose of this class is to instantiate the 1DS logger which allows events to be sent from adaptive cards 
+ */
+export class Microsoft1DSLogger {
```
[the remark this answers is not in the document]
Comment (utsavlathia) [reply source/nodejs/adaptivecards/src/logging/Microsoft1DSLogger/Microsoft1DSLogger.ts]: Yes we are having an abstract base Logger class. @choupete has created this and will be putting his PR in soon.

[the remark this answers is not in the document]
Comment (utsavlathia) [reply source/nodejs/adaptivecards/src/logging/Microsoft1DSLogger/Microsoft1DSLogger.ts]: I added the interface and renamed the file in my last commit to make it more clear

File: source/nodejs/adaptivecards/package.json (hunk 4 above)
[the remark this answers is not in the document]
Comment (utsavlathia) [reply source/nodejs/adaptivecards/package.json]: For the dependency, this has now been changed to be optional

[the remark this answers is not in the document]
Comment (dclaux) [reply source/nodejs/adaptivecards/package.json]: I really think the dependency, optional or not, needs to be completely removed from the adaptivecards package/code. The dependency should be in the other direction, e.g. we need a 1ds package that has a dependency on the adaptivecards package.

File: source/nodejs/tsconfig.json [hunk 8]
```diff
@@ -1,10 +1,10 @@
 {
 	"compilerOptions": {
 		"module": "commonjs",
-		"target": "es5",
+		"target": "es2017",
 		"lib": [
 			"dom",
-			"es2015"
+			"es2017"
```
Comment (shalinijoshi19) [inline source/nodejs/tsconfig.json]: was there a reason we needed to upgrade here? I am all for it but just want to make sure this will not break existing assumptions/code/users etc.

Comment (utsavlathia) [reply source/nodejs/tsconfig.json]: Previously it was to use the object.entries function to simply be able to go through key value pairs in an object. Now I have switched the code to do this in a easier fashion with an explicit for loop, so this upgrade is no longer necessary

**Comment by paulcam206** [comment]
closing out old PRs
